build/configs/common_download.sh

```diff
@@ -0,0 +1,390 @@
+	fi
+	echo "============================="
+	exe_name=$(get_executable_name ${parts[$partidx]})
+	board_download $TTYDEV ${offsets[$partidx]} ${exe_name} ${sizes[partidx]}
+
+	echo ""
+	echo "Download $exe_name COMPLETE!"
+}
+
+download_all()
+{
+	echo "Starting Download..."
+	found_kernel=false
+
+	for partidx in ${!parts[@]}; do
+
+		if [[ "${CONFIG_APP_BINARY_SEPARATION}" != "y" ]];then
+			if [[ "${parts[$partidx]}" == "userfs" ]];then
+				continue
+			fi
+		fi
+
+		if [[ "${parts[$partidx]}" == "kernel" ]];then
+			if [[ $found_kernel == true ]];then
+				continue
+			fi
+			found_kernel=true
+		fi
+
+		if [[ "${parts[$partidx]}" == "ftl" ]];then
+			continue
+		fi
+
+		if [[ "${parts[$partidx]}" == "config" ]];then
+			continue
+		fi
+
+		exe_name=$(get_executable_name ${parts[$partidx]})
+		[ "No Binary Match" = "${exe_name}" ] && continue
+
+		echo ""
+		echo "=========================="
+		echo "Downloading ${parts[$partidx]} binary"
+		echo "=========================="
+
+		board_download $TTYDEV ${offsets[$partidx]} ${exe_name} ${sizes[partidx]}
+	done
+	echo ""
+	echo "Download COMPLETE!"
+}
+
+erase()
+{
+	echo "Starting Erase..."
+	if [[ $2 == "all" || $2 == "ALL" ]];then
+		echo ""
+		echo "=========================="
+		echo "      Erasing All"
+		echo "=========================="
+		for partidx in ${!parts[@]}; do
+			if [[ "${parts[$partidx]}" == "userfs" ]];then
+				continue
+			elif [[ "${parts[$partidx]}" == "ss" ]];then
+				continue
+			else
+				echo ""
+				echo "=========================="
+				echo "Erasing ${parts[$partidx]} partition"
+				echo "=========================="
+			fi
+
+			board_erase $TTYDEV ${offsets[$partidx]} ${sizes[partidx]}
+		done
+	else
+		found_kernel=false
+		for partidx in ${!parts[@]}; do
+			if [[ $2 == "kernel" || $2 == "KERNEL" ]];then
+				if [[ "${parts[$partidx]}" != "kernel" ]];then
+					continue
+				elif [[ $found_kernel == false ]];then
+					found_kernel=true
+				else
+					continue
+				fi
+			elif [[ $2 == "ota" || $2 == "OTA" ]];then
+				if [[ "${parts[$partidx]}" != "kernel" ]];then
+					continue
+				elif [[ $found_kernel == false ]];then
+					found_kernel=true
+					continue
+				else
+					flash_ota=false
+				fi
+			elif [[ $2 == "ss" || $2 == "SS" ]];then
+				if [[ "${parts[$partidx]}" != "ss" ]];then
+					continue
+				fi
+			elif [[ $2 == "userfs" || $2 == "USERFS" ]];then
+				if [[ "${parts[$partidx]}" != "userfs" ]];then
+					continue
+				fi
+			else
+				printf "\n## Invalid Option ##\n\n"
+				dwld_help
+				break
+			fi
+
+			echo ""
+			echo "=========================="
+			echo "Erasing ${parts[$partidx]} partition"
+			echo "=========================="
+			board_erase $TTYDEV ${offsets[$partidx]} ${sizes[partidx]}
+		done
+	fi
+	echo ""
+	echo "Erase COMPLETE!"
+}
+
+# Start here
+
+echo -e $WARNING
+
+cmd_args=$(echo $@ | tr '[:upper:]' '[:lower:]')
+
+# Treat adding the USB rule first
+usb_rule ${cmd_args[@]};
```

Comment:
Do we need to have a separate API called usb_rule? Can we move this to pre_download?

Reply:
Okay, will move usb_rule to pre download.

Reply:
Done.